Objects/unionobject.c

```diff
@@ -0,0 +1,433 @@
+                PyObject* nested_arg = PyTuple_GET_ITEM(nested_args, j);
+                Py_INCREF(nested_arg);
+                PyTuple_SET_ITEM(flattened_args, pos, nested_arg);
+                pos++;
+            }
+        } else {
+            Py_INCREF(arg);
+            PyTuple_SET_ITEM(flattened_args, pos, arg);
+            pos++;
+        }
+    }
+    Py_INCREF(flattened_args);
+    return flattened_args;
+}
+
+static PyObject*
+dedup_and_flatten_args(PyObject* args)
+{
+    args = flatten_args(args);
+    int arg_length = PyTuple_GET_SIZE(args);
+    PyObject* temp[arg_length];
+    for (int i = 0; i < arg_length ; i++) { temp[i] = NULL; }
+
+    // Add unique elements to an array.
+    int temp_count = 0;
+    for(int i = 0; i < arg_length; i++) {
+        int is_duplicate = 0;
+        PyObject* i_tuple = PyTuple_GET_ITEM(args, i);
+        for(int j = i + 1; j < arg_length; j++) {
+            PyObject* j_tuple = PyTuple_GET_ITEM(args, j);
+            if (i_tuple == j_tuple) is_duplicate = 1;
+        }
+        if (!is_duplicate) {
+            Py_INCREF(i_tuple);
+            temp[temp_count] = i_tuple;
+            temp_count++;
+        }
+    }
+
+    // Create new tuple from deduplicated array.
+    PyObject* temp_arg = NULL;
+    PyObject* new_args = PyTuple_New(temp_count);
+    for(int k = 0; k < arg_length; k++) {
+        temp_arg = temp[k];
+
+        if (temp_arg != NULL)
+            PyTuple_SET_ITEM(new_args, k, temp_arg);
+    }
+    Py_INCREF(new_args);
+    return new_args;
+}
+
+static int
+is_typevar(PyObject *obj)
+{
+    return is_typing_name(obj, "TypeVar");
+}
+
+static int
+is_new_type(PyObject *obj)
+{
+    if (!PyObject_IsInstance(obj, (PyObject *)&PyFunction_Type)) {
+        return 0;
+    }
+    PyObject *module = PyObject_GetAttrString(obj, "__module__");
+    if (module == NULL) {
+        return 0;
+    }
+    return PyUnicode_Check(module) && _PyUnicode_EqualToASCIIString(module, "typing");
+}
+
+static int
+is_unionable(PyObject *obj)
+{
+    if (obj == Py_None) {
+        return 1;
+    }
+    return (
+        is_typevar(obj) ||
+        is_new_type(obj) ||
+        PyType_Check(obj) ||
+        (PyObject_IsInstance(obj, (PyObject*)&Py_GenericAliasType)) ||
+        (PyObject_IsInstance(obj, (PyObject *)&PyType_Type) == 1) ||
+        (PyObject_IsInstance(obj, (PyObject *)&Py_UnionType) == 1));
+}
+
+static PyObject *
+union_new(PyTypeObject* self, PyObject* param)
+{
+    // Check param is a PyType or GenericAlias
+    if ((param == NULL) || !is_unionable((PyObject *)param) || !is_unionable((PyObject*)self))
+    {
+        Py_INCREF(Py_NotImplemented);
+        return Py_NotImplemented;
+    }
+
+    PyObject *param_type = param == Py_None ? (PyObject *)Py_TYPE(param) : param;
+    PyTypeObject *self_type = (PyObject *)self == Py_None ? Py_TYPE(self) : self;
+    PyObject *tuple = PyTuple_Pack(2, self_type, param_type);
+    PyObject *new_union = Py_Union(tuple);
```

Comment:
`Py_Union` will (should) return `NULL` on failure, so you need to handle that case